File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('.\n');
input.pop();
let answer = [];
input.forEach((item) => {
    const stack = [];
    let flag = 0;
    item.split('').forEach((s) => {
        if (s === '(' || s === '[') stack.push(s);
        else if ((s === ')' && stack.at(-1) === '(') || (s === ']' && stack.at(-1) === '[')) stack.pop();
        else if (
            (s === ']' && (stack.length === 0 || stack.at(-1) !== '[')) ||
            (s === ')' && (stack.length === 0 || stack.at(-1) !== '('))
        )
            flag = 1;
    });
    stack.length !== 0 || flag ? answer.push('no') : answer.push('yes');
});
console.log(answer.join('\n'));

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
using namespace std;

int main(void){
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  while(true){
    string str;
    getline(cin, str);
    if(str == ".") break;
    stack<char> s;
    int flag = false;
    for(auto c : str){
      if(c == '(' || c == '[') s.push(c);
      else if(c == ')'){
        if(s.empty() || s.top() != '('){
          flag = true;
          break;
        }
        s.pop();
      }else if(c == ']'){
        if(s.empty() || s.top() != '['){
          flag = true;
          break;
        }
        s.pop();
      }
    }
    cout << (s.empty() && !flag? "yes" : "no") << '\n';
  }
}